fleetspeak/src/e2etesting/lib/setup_components.go

```diff
@@ -0,0 +1,296 @@
+	grpcservice_pb "github.com/google/fleetspeak/fleetspeak/src/server/grpcservice/proto/fleetspeak_grpcservice"
+	services_pb "github.com/google/fleetspeak/fleetspeak/src/server/proto/fleetspeak_server"
+	"google.golang.org/grpc"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"sync"
+	"syscall"
+	"time"
+)
+
+var (
+	mysql_database = flag.String("mysql_database", "", "MySQL database name to use")
+	mysql_username = flag.String("mysql_username", "", "MySQL username to use")
+	mysql_password = flag.String("mysql_password", "", "MySQL password to use")
+)
+
+type startedProcessesPids struct {
+	server_pid  chan int
+	client_pid  chan int
+	service_pid chan int
+}
+
+func (sp *startedProcessesPids) init() {
+	sp.server_pid = make(chan int, 1)
+	sp.client_pid = make(chan int, 1)
+	sp.service_pid = make(chan int, 1)
+}
+
+func killProcess(pid int) {
+	syscall.Kill(pid, syscall.SIGTERM)
+	for {
+		process, err := os.FindProcess(pid)
+		if err != nil {
+			continue
+		}
+		err = process.Signal(syscall.Signal(0))
+	}
+}
+
+func (sp *startedProcessesPids) killAll() {
+	pids := make([]int, 0)
+	close(sp.server_pid)
+	close(sp.client_pid)
+	close(sp.service_pid)
+	for pid := range sp.server_pid {
+		pids = append(pids, pid)
+	}
+	for pid := range sp.client_pid {
+		pids = append(pids, pid)
+	}
+	for pid := range sp.service_pid {
+		pids = append(pids, pid)
+	}
+
+	for _, pid := range pids {
+		syscall.Kill(pid, syscall.SIGTERM)
+	}
+	for {
+		finished := true
+		for _, pid := range pids {
+			process, err := os.FindProcess(pid)
+			if err != nil {
+				continue
+			}
+			err = process.Signal(syscall.Signal(0))
+			if err == nil {
+				finished = false
+				break
+			} else {
+				fmt.Println(pid, err)
+			}
+		}
+		if finished {
+			break
+		}
+		time.Sleep(time.Second)
+	}
+}
+
+// Starts a command and redirects its output to main stdout
+func startProcess(cmd *exec.Cmd, pid_chan chan int) {
+	CopyAndCapture := func(w io.Writer, r io.Reader) ([]byte, error) {
+		var out []byte
+		buf := make([]byte, 1024, 1024)
+		for {
+			n, err := r.Read(buf[:])
+			if n > 0 {
+				d := buf[:n]
+				out = append(out, d...)
+				_, err := w.Write(d)
+				if err != nil {
+					return out, err
+				}
+			}
+			if err != nil {
+				// Read returns io.EOF at the end of file, which is not an error for us
+				if err == io.EOF {
+					err = nil
+				}
+				return out, err
+			}
+		}
+	}
+
+	stdoutIn, _ := cmd.StdoutPipe()
+	stderrIn, _ := cmd.StderrPipe()
+	err := cmd.Start()
+	if err != nil {
+		fmt.Println("cmd.Start() failed: " + err.Error())
+	}
+	pid_chan <- cmd.Process.Pid
+
+	// cmd.Wait() should be called only after we finish reading
+	// from stdoutIn and stderrIn.
+	// wg ensures that we finish
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		CopyAndCapture(os.Stdout, stdoutIn)
+		wg.Done()
+	}()
+
+	CopyAndCapture(os.Stderr, stderrIn)
+	wg.Wait()
+	cmd.Wait()
+}
+
+func getNewClientId(admin_port int, start_time time.Time) string {
```

Comment:
This function prints too much errors, instead of returning them. This makes code non-reusable.

What we usually want as a function structure:

fun foo() {
  if (fail)
   return error
  if (fail2)
   return error
  if (fail3)
   return error
  return success

What we don't want, since such function can't be reused.

fun bar() {
 if (fail1) fmt.Printf(...)
 if (fail2) fmt.Printf(...)
 if (fail2) fmt.Printf(...)
 return sucess
}

I'd suggest to split this function into 2:

// Gets the client id, returns an error on failure or if there is no new client.
getNewClientId(....) string, error

// Contains the retry loop, calling getNewClientId repeatedly
// The printing of errors happens in this function, but once only.
waitForNewClientId(...) string, error